### PR TITLE
Source node

### PR DESCRIPTION
**Pull request by Depau**
Implement source nodes.
See tests and docstrings for examples.

### CODE REVIEW

**Comment by Depaulicious** [comment]
Actually this won't work if there are no input files. I need to figure out a way to include them [this way](https://stackoverflow.com/a/18141519/1124621) if needed.

File: ffmpeg/_ffmpeg.py [hunk 1]
```diff
@@ -24,6 +24,27 @@ def input(filename, **kwargs):
     return InputNode(input.__name__, kwargs=kwargs).stream()
 
 
+
+def source_multi_output(filter_name, *args, **kwargs):
```
Comment (kkroening) [inline ffmpeg/_ffmpeg.py]: Are there any multi-output sources that you know of? If not maybe we should drop `source_multi_output`.

Comment (kkroening) [reply ffmpeg/_ffmpeg.py]: In any case if there is it'd be good to add an example in the docstring of a source that has multiple outputs.

File: ffmpeg/_run.py [hunk 2]
```diff
@@ -118,10 +118,11 @@ def get_args(stream_spec, overwrite_output=False):
     input_nodes = [node for node in sorted_nodes if isinstance(node, InputNode)]
     output_nodes = [node for node in sorted_nodes if isinstance(node, OutputNode)]
     global_nodes = [node for node in sorted_nodes if isinstance(node, GlobalNode)]
-    filter_nodes = [node for node in sorted_nodes if isinstance(node, FilterNode)]
+    filter_nodes = [node for node in sorted_nodes if isinstance(node, (FilterNode, SourceNode))]
```
Comment (kkroening) [inline ffmpeg/_run.py]: How about we just check if it's a `FilterableNode` instead of checking for both implementations?

File: ffmpeg/_run.py [hunk 3]
```diff
@@ -118,10 +118,11 @@ def get_args(stream_spec, overwrite_output=False):
...
     stream_name_map = {(node, None): _get_stream_name(i) for i, node in enumerate(input_nodes)}
     filter_arg = _get_filter_arg(filter_nodes, outgoing_edge_maps, stream_name_map)
-    args += reduce(operator.add, [_get_input_args(node) for node in input_nodes])
+    if len(input_nodes) > 0:
+        args += reduce(operator.add, [_get_input_args(node) for node in input_nodes])
```
Comment (kkroening) [inline ffmpeg/_run.py]:
```diff
-    if len(input_nodes) > 0:
-        args += reduce(operator.add, [_get_input_args(node) for node in input_nodes])
+    args += reduce(operator.add, [_get_input_args(node) for node in input_nodes], [])
```

File: ffmpeg/_ffmpeg.py (hunk 1 above)
Comment (Depaulicious) [reply ffmpeg/_ffmpeg.py]: Actually I have no idea, I just added it for completeness. 👍 for docstring and tests.
Later this week or next week.

File: ffmpeg/_run.py (hunk 2 above)
Comment (Depaulicious) [reply ffmpeg/_run.py]: IDK why I didn't think of that. Anyway, to obtain the inputs using the libavfilter syntax I pointed out in the stackoverflow link, we would need to do some more checks anyway.

**Comment by kkroening** [comment]
I'll try to get this merged soon and add some examples. I think it could be really useful.

Thanks again for implementing this, and sorry I haven't gotten around to merging it yet.

**Comment by CNugteren** [comment]
@kkroening Any updates on this? It would be great to create new streams from input different than from reading in from file, e.g. using the `color` filter (https://ffmpeg.org/ffmpeg-filters.html#Examples-114). My use-case would be to add a few seconds of black at the start and at the end of the video, something which isn't currently possible with `ffmpeg-python` it seems.

**Comment by abingham** [comment]
Any plans to merge this?

**Comment by silviot** [comment]
@Depau @kkroening I needed this feature and rebased this branch on current master. Feel free to update this PR.

https://github.com/silviot/ffmpeg-python/commits/source-node

**Comment by Depau** [comment]
Replaced by #418

**Pull request closed without merging**